### PR TITLE
Label-driven recurring job

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/issues/467

### CODE REVIEW

File: manager/volume.go
Comment (innobead) [inline manager/volume.go:857]: Check if it exists before deletion.

File: manager/volume.go
Comment (innobead) [inline manager/volume.go:794]: Check if it does not exist before addition.

File: controller/recurring_job_controller.go
Comment (innobead) [inline controller/recurring_job_controller.go:101]: "Staring ....". Encourage use verb+ing/ed in the log.

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Just few comments required to resolve.

**Comment by innobead** [comment]
Also need @shuo-wu or @joshimoo (one of maintainers) to review before merging.

File: csi/controller_server.go
Comment (c3y1huang) [reply csi/controller_server.go:191]: https://github.com/longhorn/longhorn/issues/2865

File: manager/volume.go
Comment (c3y1huang) [reply manager/volume.go:857]: Is this necessary?
[The delete built-in function deletes the element with the specified key (m[key]) from the map. If m is nil or there is no such element, delete is a no-op.](https://pkg.go.dev/builtin#delete)

Comment (c3y1huang) [reply manager/volume.go:857]: Thanks for @innobead explanation. The comment is referring to the following `UpdateVolume`.

File: app/recurring_job.go
Comment (c3y1huang) [reply app/recurring_job.go]: Fixed.

File: manager/volume.go
Comment (c3y1huang) [reply manager/volume.go]: Fixed.

File: manager/volume.go
Comment (c3y1huang) [reply manager/volume.go]: Fixed

File: controller/recurring_job_controller.go
Comment (c3y1huang) [reply controller/recurring_job_controller.go]: Fixed.

File: upgrade/v110to120/upgrade.go
Comment (shuo-wu) [inline upgrade/v110to120/upgrade.go:294]: Do the old cron jobs have any dependencies?

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: types/resource.go
Comment (c3y1huang) [reply types/resource.go:174]: Addressed.

File: types/deepcopy.go
Comment (c3y1huang) [reply types/deepcopy.go]: Fixed.

File: datastore/kubernetes.go
Comment (c3y1huang) [reply datastore/kubernetes.go]: Renamed.

File: datastore/longhorn.go
Comment (c3y1huang) [reply datastore/longhorn.go]: Renamed.

File: datastore/longhorn.go
Comment (c3y1huang) [reply datastore/longhorn.go]: Renamed.

File: datastore/longhorn.go
Comment (c3y1huang) [reply datastore/longhorn.go]: Fixed.

File: controller/recurring_job_controller.go
Comment (c3y1huang) [reply controller/recurring_job_controller.go]: Removed.

File: controller/recurring_job_controller.go
Comment (c3y1huang) [reply controller/recurring_job_controller.go]: Removed, Not required after implement foreground deletion propagation.

File: controller/recurring_job_controller.go
Comment (c3y1huang) [reply controller/recurring_job_controller.go]: Sorry missed this one. Removed.

File: controller/recurring_job_controller.go
Comment (c3y1huang) [reply controller/recurring_job_controller.go:237]: Fixed.

File: controller/recurring_job_controller.go
Comment (c3y1huang) [reply controller/recurring_job_controller.go]: Fixed.

File: controller/recurring_job_controller.go
Comment (c3y1huang) [reply controller/recurring_job_controller.go]: Renamed.

File: app/recurring_job.go
Comment (c3y1huang) [reply app/recurring_job.go]: Fixed.

File: app/recurring_job.go
Comment (c3y1huang) [reply app/recurring_job.go:174]: Done.

File: app/recurring_job.go
Comment (c3y1huang) [reply app/recurring_job.go]: Fixed.

File: api/model.go
Comment (c3y1huang) [reply api/model.go]: Fixed.

File: api/recurringjob.go
Comment (c3y1huang) [reply api/recurringjob.go:64]: Fixed.

File: manager/recurringjob.go
Comment (c3y1huang) [reply manager/recurringjob.go]: Addressed.

File: upgrade/v110to120/upgrade.go
Comment (c3y1huang) [reply upgrade/v110to120/upgrade.go:296]: Fixed.

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: types/types.go
Comment (c3y1huang) [reply types/types.go:66]: We need to differentiate between job and job group labels. To use `recurring-job.longhorn.io/<jobName>` and `recurring-job-group.longhorn.io/<groupName>` was agreed and discussed in the LEP.

File: datastore/longhorn.go
Comment (c3y1huang) [reply datastore/longhorn.go:2089]: This is for StorageClass `recurringJobs` that are planning to be removed in the next version.

Comment (shuo-wu) [reply datastore/longhorn.go:2089]: For `recurringJobs` in StorageClass, we can handle it in the CSI plugin: create CRs then convert them to recurring job selectors.

File: types/types.go
Comment (shuo-wu) [reply types/types.go:66]: I know. I mean `longhorn.io/` typically is used as the label key prefix. Maybe you can use `longhorn.io/recurring-job-<jobName>` and `longhorn.io/recurring-job-group-<groupName>`

File: manager/recurringjob.go
Comment (c3y1huang) [reply manager/recurringjob.go:101]: Yes because I don't see much advantage atm for using map as array for this case. We could enhance this later if a use case comes up.

File: api/model.go
Comment (c3y1huang) [reply api/model.go]: Fixed.

File: api/model.go
Comment (c3y1huang) [reply api/model.go]: Fixed.

File: api/model.go
Comment (c3y1huang) [reply api/model.go]: Removed.

File: client/generated_node_input.go
Comment (c3y1huang) [reply client/generated_node_input.go]: Fixed.

File: datastore/longhorn.go
Comment (c3y1huang) [reply datastore/longhorn.go]: > 1. if `job.Concurrency == 0`, it's better to consider it as the default value. Since the old recurring jobs in volumeSpec or StorageClass don't contain this field. Or users may not care about this field at all.

Fixed.

> 2. Need to check `Task`

Fixed.

File: datastore/longhorn.go
Comment (c3y1huang) [reply datastore/longhorn.go]: Not required anymore since Longhorn no longer deal with `recurringJobs` in volume spec as discussed.

File: datastore/longhorn.go
Comment (c3y1huang) [reply datastore/longhorn.go]: Removed.

File: deploy/install/01-prerequisite/06-storageclass.yaml
Comment (c3y1huang) [reply deploy/install/01-prerequisite/06-storageclass.yaml:27]: Removed and added `recurringJobSelector` to examples.

File: controller/recurring_job_controller.go
Comment (c3y1huang) [reply controller/recurring_job_controller.go]: Removed.

File: datastore/longhorn.go
Comment (c3y1huang) [reply datastore/longhorn.go]: Removed.

File: manager/volume.go
Comment (c3y1huang) [reply manager/volume.go]: Removed.

File: manager/volume.go
Comment (c3y1huang) [reply manager/volume.go]: Done.

File: types/resource.go
Comment (c3y1huang) [reply types/resource.go]: Fixed.

File: types/resource.go
Comment (c3y1huang) [reply types/resource.go]: Addressed.

**Review by shuo-wu — APPROVED** [review]

**Review by innobead — APPROVED** [review]

**Pull request merged**
